core/src/parser/mod.rs

```diff
@@ -0,0 +1,107 @@
+pub trait Parser<'a> {
+    /// The input type provided to the parser.
+    type Input;
+
+    /// The output type of the parser.
+    type Output;
+
+    /// Parse the input and provide the output.
+    fn parse(input: Self::Input) -> Result<Self::Output, ParseError>;
+}
+
+/// The results of parsing Rust source input.
+#[derive(Default, Debug)]
+pub struct ParsedData {
+    /// Structs defined in the source
+    pub structs: Vec<RustStruct>,
+    /// Enums defined in the source
+    pub enums: Vec<RustEnum>,
+    /// Type aliases defined in the source
+    pub aliases: Vec<RustTypeAlias>,
+}
+
+impl ParsedData {
+    /// Add the parsed data from `other` to `self`.
+    pub fn add(&mut self, mut other: Self) {
+        self.structs.append(&mut other.structs);
+        self.enums.append(&mut other.enums);
+        self.aliases.append(&mut other.aliases);
+    }
+
+    fn push(&mut self, rust_thing: RustItem) {
+        match rust_thing {
+            RustItem::Struct(s) => self.structs.push(s),
+            RustItem::Enum(e) => self.enums.push(e),
+            RustItem::Alias(a) => self.aliases.push(a),
+        }
+    }
+}
+
+impl<'a> Parser<'a> for ParsedData {
+    type Input = &'a Item;
+    type Output = ParsedData;
+
+    fn parse(input: Self::Input) -> Result<Self::Output, ParseError> {
+        let mut ret = ParsedData::default();
+        if let Some(rust_item) = RustItem::parse(input)? {
+            ret.push(rust_item);
+        }
+        Ok(ret)
+    }
+}
+
+/// Parse the given Rust source string into `ParsedData`.
+pub fn parse(input: &str) -> Result<ParsedData, ParseError> {
+    // We will only produce output for files that contain the `#[typeshare]`
+    // attribute, so this is a quick and easy performance win
+    if !input.contains(TYPESHARE) {
+        return Ok(ParsedData::default());
+    }
+
+    // Parse and process the input, ensuring we parse only items marked with
+    // `#[typeshare]`
+    let source = syn::parse_file(input)?;
+
+    let iter = flatten_items(source.items.iter()).map(ParsedData::parse);
+
+    let mut ret = ParsedData::default();
+    for parsed_data in iter {
+        ret.add(parsed_data?);
+    }
+    Ok(ret)
+}
+
+impl<'a> Parser<'a> for RustItem {
+    type Input = &'a Item;
+    type Output = Option<RustItem>;
+
+    fn parse(input: Self::Input) -> Result<Self::Output, ParseError> {
+        match input {
+            syn::Item::Struct(item_struct) if has_typeshare_annotation(&item_struct.attrs) => {
+                RustStruct::parse(item_struct).map(Some)
+            }
+            syn::Item::Enum(item_enum) if has_typeshare_annotation(&item_enum.attrs) => {
+                RustEnum::parse(item_enum).map(Some)
+            }
+            syn::Item::Type(item_type) if has_typeshare_annotation(&item_type.attrs) => {
+                RustTypeAlias::parse(item_type).map(Some)
+            }
+            _ => Ok(None),
+        }
+    }
+}
```

Comment:
Could you partially revert this to be more similar to the original? My concern is mostly with this flow of `RustItem` -> `ParsedData containing a single item` -> `collect into ParsedData`. This involves a lot of unnecessary allocations, and more semantically it round trips through these types in unnecessary ways.